### PR TITLE
noise updates

### PR DESCRIPTION
**Pull request by arnetheduck**
* clear secrets explicitly
* simplify keygen
* avoid some trivial memory allocations
* fix little endian encoding of nonce

### CODE REVIEW

File: libp2p/protocols/secure/noise.nim [hunk 1]
```diff
@@ -124,13 +121,12 @@ proc encryptWithAd(state: var CipherState, ad, data: openarray[byte]): seq[byte]
   result &= tag
   trace "encryptWithAd", tag = byteutils.toHex(tag), data = result.shortLog, nonce = state.n - 1
 
-proc decryptWithAd(state: var CipherState, ad, data: openarray[byte]): seq[byte] =
+proc decryptWithAd(state: var CipherState, ad, data: openArray[byte]): seq[byte] =
   var
-    tagIn = data[^ChaChaPolyTag.len..data.high].intoChaChaPolyTag
-    tagOut = tagIn
+    tagIn = data.toOpenArray(data.len - ChaChaPolyTag.len, data.high).intoChaChaPolyTag
+    tagOut: ChaChaPolyTag
```
Comment (dryajov) [inline libp2p/protocols/secure/noise.nim:127]: where is `tagOut` being set now?

Comment (arnetheduck) [reply libp2p/protocols/secure/noise.nim:127]: `tagOut` comes from the encrypted data (it's appended as part of encryption and verified as part of decryption)

Comment (sinkingsugar) [reply libp2p/protocols/secure/noise.nim:127]: `ChaChaPoly.decrypt` is a proc(..., var _: T, ...)

**Comment by sinkingsugar** [comment]
Looks good to me @arnetheduck

File: libp2p/protocols/secure/noise.nim [hunk 2]
```diff
@@ -491,7 +510,7 @@ proc newNoise*(
     rng: rng,
     outgoing: outgoing,
     localPrivateKey: privateKey,
-    localPublicKey: privateKey.getKey().tryGet(),
+    localPublicKey: privateKey.getKey().tryGet().getBytes().tryGet(),
```
Comment (cheatfate) [inline libp2p/protocols/secure/noise.nim:513]: This sandwich looks ugly.

Comment (sinkingsugar) [reply libp2p/protocols/secure/noise.nim:513]: 🥪

File: libp2p/protocols/secure/noise.nim (hunk 1 above)
Comment (dryajov) [reply libp2p/protocols/secure/noise.nim:127]: Ugh, ok, but looks fishy, it either was wrong or it's wrong now. We used to set `tagOut = tagIn`, now we aren't...

Comment (arnetheduck) [reply libp2p/protocols/secure/noise.nim:127]: the underlying chacha it calls is `out-only` for tag

Comment (arnetheduck) [reply libp2p/protocols/secure/noise.nim:127]: I guess the chacha api could be made better but whatever

**Pull request merged**